Backport DataTables using merge commit

nalimilan opened the pull request:
This is equivalent to https://github.com/JuliaStats/DataFrames.jl/pull/1214 but using `git merge` rather than `git rebase`. The merge commit is quite small (it can be seen locally using `git show` from the head of the branch). DataTables history has been preserved, except for the DataFrames ->DataTables rename which has been erased using [this script](https://gist.github.com/nalimilan/9c7c861dfafbc3601487bd03694a0931).

nalimilan commented:
Of course this cannot be merged via the GitHub interface since that's not fast-forward. It will have to be done manually.

nalimilan commented:
(Even if the colors don't indicate it, tests are green on Travis with Julia 0.6.)

rofinn commented:
Diffing `rf/datatables` and `nl/datatables-backport` wasn't very helpful for reviewing, but a few things I did notice were:

1. A few remaining incorrect names (e.g., `dt` -> `df`, JuliaData -> JuliaStats).
2. Documentation that needs to be removed or fixed (e.g., the formulas documentation should be removed)

nalimilan commented:
Good catch! New version should fix these issues.

ararslan commented on docs/src/lib/utilities.md:
```diff
@@ -9,14 +8,13 @@ CurrentModule = DataFrames
 eltypes
 head
 completecases
-completecases!
```
Shouldn't we be removing `completecases` here in addition to `completecases!`?

ararslan replied:
nvm looks like they do different things; the former identifies complete cases as a boolean vector and the latter is now `dropnull!`.

ararslan commented on docs/src/man/reshaping_and_pivoting.md:
```diff
@@ -3,8 +3,9 @@
-iris = dataset("datasets", "iris")
+using DataFrames
+using CSV
+iris = CSV.read(joinpath(Pkg.dir("DataFrames"), "test/data/iris.csv"), DataFrame)
```
The slashes in the path here kind of defeat the purpose of using `joinpath` 😉

ararslan commented on src/DataFrames.jl:
```diff
@@ -25,38 +19,23 @@ import Base: ==, |>
        colwise,
        combine,
        completecases,
-       completecases!,
```
Same comment regarding `completecases`

ararslan commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -229,23 +226,17 @@ Base.ndims(::AbstractDataFrame) = 2
 ##############################################################################
 
+# Imported in DataFrames.jl for compatibility across Julia 0.4 and 0.5
+Base.:(==)(df1::AbstractDataFrame, df2::AbstractDataFrame) = isequal(df1, df2)
```
Didn't we change this recently in DataTables? Having `==` call `isequal` defeats the purpose of having them be separate functions. If this is only for 0.4 and 0.5 compatibility, we don't support those anymore, so this could be removed.

quinnj replied:
Yeah, I thought this looked fishy too.

nalimilan replied:
Yeah, that's also related to issues around how to define `==` for `Nullable`. We should fix this in later PRs.

ararslan commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -405,14 +378,43 @@ function StatsBase.describe(io, df::AbstractDataFrame)
     end
 end
 
+function StatsBase.describe{T}(io::IO, X::AbstractVector{Union{T, Null}})
```
This syntax is deprecated, should use `where`

ararslan commented on src/abstractdataframe/io.jl:
```diff
@@ -5,7 +5,7 @@
 
 function escapedprint(io::IO, x::Any, escapes::AbstractString)
-    print(io, x)
+    ourshowcompact(io, x)
```
Where are we defining this? It doesn't appear to be defined in this file.

ararslan replied:
Found it, it's in src/abstractdataframe/show.jl, so this is fine.

ararslan commented on src/abstractdataframe/join.jl:
```diff
@@ -2,136 +2,191 @@
-            lcount += lc
-        end
+# Like similar, but returns a array that can have nulls and is initialized with nulls
+similar_nullable{T}(dv::AbstractArray{T}, dims::Union{Int, Tuple{Vararg{Int}}}) =
```
`where`

ararslan commented on src/abstractdataframe/join.jl:
```diff
@@ -2,136 +2,191 @@
-        left_pos += lc
-        right_pos += rc
-    end
+DataFrameJoiner{DF1<:AbstractDataFrame, DF2<:AbstractDataFrame}(dfl::DF1, dfr::DF2, on::Union{Symbol,Vector{Symbol}}) =
```
`where`

ararslan commented on src/abstractdataframe/join.jl:
```diff
@@ -2,136 +2,191 @@
-        val = df[i,:]
-        if haskey(poolref, val)
-            refs[i] = poolref[val]
+function fillcolumn!{T1, T2}(dfcol::AbstractVector{T1}, refcol::AbstractVector{T2},
```
`where`

ararslan commented on src/abstractdataframe/sort.jl:
```diff
@@ -55,69 +55,36 @@ ordering(col::ColumnIndex, lt::Function, by::Function, rev::Bool, order::Orderin
 #         sort the original (presumably larger) DataFrame
 
-type DFPerm{O<:Union{Ordering, AbstractVector}, DF<:AbstractDataFrame} <: Ordering
+immutable DFPerm{O<:Union{Ordering, AbstractVector}, DF<:AbstractDataFrame} <: Ordering
```
`struct`

ararslan commented on src/dataframe/dataframe.jl:
```diff
@@ -689,7 +663,7 @@ function deleterows!(df::DataFrame, ind::AbstractVector{Int})
     iind = 1
     ikeep = 1
-    keep = Vector{Int}(n - length(ind2))
+    keep = Array{Int}(n-length(ind2))
```
?

nalimilan replied:
This was `Array` before the fork, and has been changed to `Vector` in DataFrames since then. I've changed the merge commit to use `Vector` since that's slightly better.

ararslan commented on src/other/utils.jl:
```diff
@@ -96,95 +95,54 @@ end
 #' DataFrames.gennames(10)
 function gennames(n::Integer)
-    res = Vector{Symbol}(n)
+    res = Array{Symbol}(n)
```
Another seemingly unnecessary change from `Vector` to `Array`

ararslan commented on src/other/utils.jl:
```diff
@@ -96,95 +95,54 @@ end
-#' DataFrames.countna(@data([1, 2, 3]))
-countna(da::DataArray) = sum(da.na)
+#' DataFrames.countnull([1, 2, 3])
+function countnull(a::AbstractArray)
```
Seems like this should either be defined in Nulls or deprecated in favor of `count(isnull, a)`

quinnj replied:
I think it's `countnull` to allow, e.g. `DataArray`/`NullableArray`/`CategoricalArray` to define optimized versions.

ararslan replied:
You still can quite easily:
```julia
Base.count(::typeof(isnull), a::SweetArray) = # awesome implementation
```

nalimilan replied:
Yeah, let's remove it in a subsequent PR.

ararslan commented on src/subdataframe/subdataframe.jl:
```diff
@@ -5,6 +5,42 @@
 ##
 ##############################################################################
 
+if VERSION >= v"0.6.0-dev.2643"
```
Could simplify this to remove the conditional

ararslan commented on test/data.jl:
```diff
@@ -42,18 +34,24 @@ module TestData
 
     #test_group("assign")
-    df6[3] = @data(["un", "deux", "troix", "quatre"])
+    df6[3] = ["un", "deux", "trois", "quatre"]
```
lol

nalimilan replied:
That was really a shame for all JuliaStats! ;-)

ararslan replied:
C'est dommage!

nalimilan commented on src/dataframerow/dataframerow.jl:
```diff
@@ -37,49 +37,76 @@ Base.done(r::DataFrameRow, s) = s > length(r)
+    (ncol(r1.df) == ncol(r2.df)) ||
+        throw(ArgumentError("Rows of the data tables that have different number of columns cannot be compared ($(ncol(df1)) and $(ncol(df2)))"))
+    @inbounds for i in 1:ncol(r1.df)
+        isless(r1.df[i][r1.row], r2.df[i][r2.row]) && return true
```
Since we're doing a general review, we should have a look at the semantics of this function at some point. It was added in https://github.com/JuliaData/DataTables.jl/pull/17 and I'm not sure what's its purpose. Its behavior looks quite counter-intuitive to me.

For now I've kept it, but simplified it a bit since `null` uses the right `isless` definition.

nalimilan replied:
See https://github.com/JuliaStats/DataFrames.jl/pull/1222. Actually, the way the code was written looked weird to me, but the behavior was OK.

nalimilan commented:
I've fixed the issues which are related to the merge. Let's handle the others later (using Femtocleaner where possible). I've also noticed a few remaining uses of `unsafe_get`, I've removed them from @quinnj's commit to keep history clean.

I've just pushed it to master since I'm afraid of what GitHub is going to do in this complex situation. We should have a look at the failures on Julia 0.7, they don't seem too hard to fix.

quinnj commented:
Note that one of the 0.7 failures is a bug in current isbits Union array copying. A fix will be incoming to Base in the next few days.
